cpp/arcticdb/column_store/column.cpp

```diff
@@ -220,66 +220,65 @@ void Column::append(const Column& other, position_t at_row) {
+    size_t physical_rows = row_count();
+    auto& buffer = data_.buffer();
+
+    util::check(sorted_pos.size() == physical_rows, "Mismatch between sorted_pos size and row_count: {} != {}",
+        sorted_pos.size(), physical_rows);
+
+    type().visit_tag([&buffer, &sorted_pos, &physical_rows] (auto tdt) {
+        using TagType = decltype(tdt);
+        using RawType = typename TagType::DataTypeTag::raw_type;
+
+        for (auto i=0u; i<physical_rows; ++i){
+            if (i != sorted_pos[i]){
+                auto& current = buffer.cast<RawType>(i);
+                // Amortized O(1) complexity, because each iteration places an element where it's supposed to go
+                // and once an element is in it's sorted position we never move it.
+                while (i != sorted_pos[i]){
+                    auto move_to = sorted_pos[i];
+                    std::swap(sorted_pos[i], sorted_pos[move_to]);
+                    std::swap(current, buffer.cast<RawType>(move_to));
+                }
+            }
+        }
+    });
+}
+
 void Column::sort_external(const JiveTable& jive_table) {
     auto rows = row_count();
     if(!is_sparse()) {
-        auto unsorted = jive_table.unsorted_rows_;
-        auto& buffer = data_.buffer();
-        type().visit_tag([&jive_table, &buffer, &unsorted] (auto tdt) {
-            using TagType = decltype(tdt);
-            using RawType = typename TagType::DataTypeTag::raw_type;
-
-            auto loc = unsorted.get_first();
-            auto tmp =  buffer.cast<RawType>(jive_table.orig_pos_[loc]);
-            for(auto i = 0u; i < jive_table.num_unsorted_; ++i) {
-                std::swap(tmp, buffer.cast<RawType>(loc));
-                unsorted.set(loc, false);
-                const auto next_pos = jive_table.sorted_pos_[loc];
-                if(unsorted[next_pos]) {
-                    loc = next_pos;
-                } else {
-                   loc =  unsorted.get_first();
-                   tmp = buffer.cast<RawType>(jive_table.orig_pos_[loc]);
-                }
-            }
-            util::check(!unsorted.any(), "Did not sort all possible values, still have {} unsorted", unsorted.count());
-        });
+        auto sorted_pos = jive_table.sorted_pos_;
+        physical_sort_external(std::move(sorted_pos));
     } else {
         const auto& sm = sparse_map();
-
         bm::bvector<>::enumerator en = sm.first();
-        bm::bvector<>::enumerator en_end = sm.end();
         util::BitMagic new_map;
-
-        while (en < en_end) {
-            auto bv_index = *en;
-            new_map.set(jive_table.sorted_pos_[bv_index]);
-            ++en;
+        // The additional allocation is of the same size as the jive table
+        // and is needed for a significant speed improvement.
+        // We could instead use a std::map and sacrifice some speed for smaller allocations.
+        auto sorted_logical_to_physical = std::vector<uint32_t>(jive_table.sorted_pos_.size());
```

Comment:
I don't know how large is the jive table, if it's large this can have some overhead because the vector will call the default constructor of uint32_t setting everything to 0. Since we're not changing the size of the vector maybe a unique_ptr<uinr32_t[]> would be better. It the jive table is small this doesn't matter at all.

Reply:
The jive table is the size of a column, but I think since there is likely to be only one jive table per segment we can handle this extra allocation. Ultimately it would be nice to allocated it as a buffer of regular-sized blocks from the slab allocator, but for that we need better iterator that can handle the contiguous memory vs evenly-sized blocks vs irregular blocks without branching on each iteration

Reply:
A caveat with this allocation is that currently we are running the sort_external sequentially on each column of the output.
Also for this allocation we don't care what are the initial values for the allocated memory (because we set them afterwards).

So, we could also do exactly one allocation for the entire sort and reuse the memory in different calls of sort_external.
However I decided against that because this will block running the sort_external calls in parallel, which we could currently do (with the cost of extra memory).

Reply:
Generally speaking we parallelize over segments, not over columns, so it's safe to assume that we only need to allocate the memory once.

Reply:
Sounds good, I did the change. 
This will also likely improve the speed in case of many very sparse columns on which we call sort_external, for which I don't have a benchmark.
I reran the other benchmarks and results were unchanged.